### PR TITLE
Chain iterators

### PR DESCRIPTION
**Pull request by baloo**
Allow to write something like:
``` rust
let first = db.search("foo");
let second = db.search("bar");
first.chain(second);
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -855,6 +867,59 @@ impl<I, F> StreamingIterator for TakeWhile<I, F>
...
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        if !self.done {
+            self.it.advance();
+            if self.it.get().is_none() {
+                self.done = true;
+                self.then.advance();
```
Comment (baloo) [inline src/lib.rs:1376]: I'm not quite sure this is safe. What if `advance` has already been called. We will miss one element.

I'm not sure what I'm supposed to do though.

File: src/lib.rs [hunk 2]
```diff
@@ -855,6 +867,59 @@ impl<I, F> StreamingIterator for TakeWhile<I, F>
...
+impl<I, J> StreamingIterator for AndThen<I, J>
+    where I: StreamingIterator,
+          J: StreamingIterator<Item=I::Item>
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        if !self.done {
+            self.it.advance();
```
Comment (sfackler) [inline src/lib.rs:1373]: The behavior of calling `advance` after the end of the first iterator is unspecified. You can either fuse `it` or do something like what the standard library does: https://doc.rust-lang.org/src/core/iter/mod.rs.html#521-525

**Comment by sfackler** [comment]
Oops, sorry for missing this!

**Comment by polarathene** [comment]
@baloo @sfackler Any chance of seeing this finished? Is it meant to be like [chain()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.chain)?

**Comment by baloo** [comment]
@polarathene If you want to take over this pull request, feel free to do :).

**Comment by sfackler** [comment]
I do agree that it should be named `chain`.

**Comment by polarathene** [comment]
@baloo I would not know what to do sorry, fairly new to Rust. I was pointed to this crate for handling my iteration logic.

**Comment by Palladinium** [comment]
@baloo I can take over this PR if you want, I'm using streaming iterators fairly extensively in my current project so I'll be re-implementing a lot of the standard library adapters as I need them.

**Comment by baloo** [comment]
@Palladinium be my guest

**Comment by Palladinium** [comment]
Implemented in #15

**Pull request closed without merging**